### PR TITLE
Define Extension Field for Mersenne 31

### PR DESCRIPTION
**Pull request by LKY-stephen**
- [x] Define General API of Extension Field
- [x] Define extension field for Mersenne 31
- [x] Tests

What has been Added:
1. Define trait Frobenius, OEF, and OptimalExtendable according to Plonky2 (mod.rs)
2. Define QuadraticOef and CubicOef with default implementation and distribution trait (quadratic.rs, cubic.rs)
3. Define quadratic and cubic extension for Mersenne31Complex (For simplicity, I used w=(i+2) for quadratic extension and w=5i for cubic extension) (extension.rs)
4. Copy two tests from Plonky2 and synthesis them in field_testing and update tests for all defined fields

Question:
Do we need another optional trait for Frobenuis? Created a new issue https://github.com/Plonky3/Plonky3/issues/90

### CODE REVIEW

**Comment by dlubarov** [comment]
Hm, another complication is, how should we handle traits such as `TwoAdicField`? I guess we could just always have `QuadraticOef` implement it, with params specified in `OptimallyExtendable`, similar to Plonky2. But I think we'll end up with other optional traits, like a `ComplexExtensionField` for irreducible `i^2 + 1`. Should we have other traits on top of `OptimallyExtendable` for such optional things?

cc @unzvfu @wborgeaud who might also have thoughts on the design

File: field/src/extension/cubic.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    fn from_wrapped_u32(n: u32) -> Self {
+        F::from_wrapped_u32(n).into()
+    }
+
+    fn from_wrapped_u64(n: u64) -> Self {
+        F::from_wrapped_u64(n).into()
+    }
+
+    fn multiplicative_group_generator() -> Self {
+        F::multiplicative_group_generator().into()
```
Comment (dlubarov) [inline field/src/extension/cubic.rs:83]: I think we want a generator of the whole EF's multiplicative group, with order `|F|^2 - 1`. Probably would need to be specified in `Extendable` as in Plonky2.

File: field/src/field.rs [hunk 2]
```diff
@@ -143,6 +144,12 @@ pub trait Field:
         }
         res
     }
+
+    // The frobenius operation to a field, for standard field, it returns itself.
+    // need to rewrite for complex field
+    fn base_frobenius(&self) -> Self {
+        *self
```
Comment (dlubarov) [inline field/src/field.rs:151]: I don't think we should provide a default impl, could be a bit dangerous as it'd be easy to forget to override it.

Comment (LKY-stephen) [reply field/src/field.rs:151]: I see. currently, I am thinking of dropping this thing due to the complex extension field is indeed not OEF, and its inverse, if using Frobenius, will have a different 'r' compare to the current implementation. The whole OEF trait is useful only when it is built upon prime field assumption. For other fields, should we have another more general trait and use algorithms  like 11.3.6 in the book <Handbook And Hyperelliptic Curve Cryptography>

**Comment by LKY-stephen** [comment]
>Could we rename something like Extendable -> OptimallyExtendable, and QuadraticExtension -> QuadraticOef?
Sure, I think this is reasonable
>  But I think we'll end up with other optional traits, like a ComplexExtensionField for irreducible i^2 + 1. 
Yes, I am also thinking about this, a quadratic oef should be a special case of quadratic extension and use the Frobenius-based algorithm.

@dlubarov thank you for checking.  Currently, I marked this PR as WIP so we can discuss some general design ideas. Once everything is settled, I will mark it ready for review.

File: field/src/extension/cubic.rs (hunk 1 above)
Comment (LKY-stephen) [reply field/src/extension/cubic.rs]: fixed

**Comment by dlubarov** [comment]
@wborgeaud would you mind taking a look when you have time later?

I guess a remaining question is the API for Frobenius (it's commented out in this branch currently). Would it make sense to have a `fn frobenius()` in `Field` itself? I think arkworks does that.

File: field/src/field.rs [hunk 3]
```diff
@@ -195,7 +196,21 @@ pub trait AbstractExtensionField<Base>:
...
+
+    fn scalar_mul(&self, scalar: Base) -> Self {
+        let value: &[Base] = self.as_base_slice();
+        let mut res = Vec::with_capacity(Self::D);
+
+        value.iter().for_each(|x| {
+            res.push(*x * scalar);
+        });
+        Self::from_base_slice(res.as_slice())
+    }
```
Comment (wborgeaud) [inline field/src/field.rs:212]: This implementation using a `Vec` seems a bit inefficient when the extension field is of the form `EF([Base; D])`, like in the case of `QuadraticOef` and `CubicOef`. 
We could overwrite it in these cases, like
```rust
fn scalar_mul(&self, scalar: Base) -> Self {
        let mut res = self.0;
        res.iter_mut().for_each(|x| {
            *x *= scalar;
        });
        Self(res)
}
```

File: goldilocks/src/lib.rs [hunk 4]
```diff
@@ -484,17 +482,8 @@ mod tests {
...
-    fn two_adic_subgroup_zerofier() {
-        test_two_adic_subgroup_zerofier::<Goldilocks>();
-    }
-
-    #[test]
-    fn two_adic_coset_zerofier() {
-        test_two_adic_coset_zerofier::<Goldilocks>();
+    fn common_tests() {
+        field_test::<Goldilocks>();
+        two_adic_field_test::<Goldilocks>();
```
Comment (wborgeaud) [inline goldilocks/src/lib.rs:487]: nit: Same as for BabyBear, is there a reason you think batching these tests is better?

File: mersenne-31/src/extension.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    type Extension = QuadraticOef<Self>;
+
+    // Verifiable in Sage with
+    // `p = 2**31 - 1  # Mersenne31`
+    // `F = GF(p)  # The base field GF(p)`
+    // `R = F['x']  # The polynomial ring over F`
+    // `K = F.extension(R.gen()**2 + 1, 'i')  # The extension field of complex`
+    // `R2 = K['y']`
+    // `f2 = R2.gen()**2-K('i')-2   # polynomial y^2-i-2`
+    // `f.is_irreducible()`
```
Comment (wborgeaud) [inline mersenne-31/src/extension.rs:17]:
```suggestion
    // Verifiable in Sage with
    // ```sage
    // p = 2**31 - 1  # Mersenne31
    // F = GF(p)  # The base field GF(p)
    // R.<x> = F[]  # The polynomial ring over F
    // K.<i> = F.extension(x^2 + 1)  # The complex extension field
    // R2.<y> = K[]
    // f2 = y^2 - i - 2 
    // assert f2.is_irreducible()
    // ```
```

File: mersenne-31/src/extension.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    const W: Self = Self::new(Mersenne31::new(2), Mersenne31::new(1));
+
+    // DTH_ROOT = W^((p - 1)/2)
+    const DTH_ROOT: Self = Self::new(Mersenne31::new(21189756), Mersenne31::new(42379512));
+
+    // Verifiable in Sage with
+    //  `g = K2('j')+6`
+    //  `for f,i in list(factor(p**4-1)):`
+    //  `  print("j to ",Integer(f)," : ",g**Integer((p^4-1)/f))`
```
Comment (wborgeaud) [inline mersenne-31/src/extension.rs:26]:
```suggestion
    // Verifiable in Sage with
    // ```sage
    // K2.<j> = K.extension(f2)
    //  g = j + 6
    // for f in factor(p^4 - 1):
    //   assert (j+6)^((p^4-1) // f) != 1 
    // ```
```

File: mersenne-31/src/extension.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    type Extension = CubicOef<Self>;
+
+    // Verifiable in Sage with
+    // `p = 2**31 - 1  # Mersenne31`
+    // `F = GF(p)  # The base field GF(p)`
+    // `R = F['x']  # The polynomial ring over F`
+    // `K = F.extension(R.gen()**2 + 1, 'i')  # The extension field of complex`
+    // `R2 = K['y']`
+    // `f2 = R2.gen()**3-5*K('i')   # polynomial y^3-5i`
+    // `f.is_irreducible()`
```
Comment (wborgeaud) [inline mersenne-31/src/extension.rs:45]: Can be cleaned similarly as above

File: mersenne-31/src/extension.rs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    const W: Self = Self::new(Mersenne31::new(0), Mersenne31::new(5));
+
+    // DTH_ROOT = W^((p - 1)/3)
+    const DTH_ROOT: Self = Self::new(Mersenne31::new(634005912), Mersenne31::new(0));
+
+    // Verifiable in Sage with
+    //  `g = K2('j')+5`
+    //  `for f,i in list(factor(p**6-1)):`
+    //  `  print("j to ",Integer(f)," : ",g**Integer((p^6-1)/f))`
```
Comment (wborgeaud) [inline mersenne-31/src/extension.rs:54]: Can be cleaned similarly as above

**Review by wborgeaud — COMMENTED** [review]
> I guess a remaining question is the API for Frobenius (it's commented out in this branch currently). Would it make sense to have a fn frobenius() in Field itself? I think arkworks does that.

I'm not sure we would need an explicit Frobenius. I think the only places where it is used in Plonky2 is to compute inverses in OEFs, but this PR gives alternative implementations that doesn't use the Frobenius. 

On the other hand, it would be pretty easy to implement it since it's the identity for prime fields. Also, since the Frobenius in the complex extension is the conjugate, we could remove the `conjugate()` function of `Mersenne31Complex` which could be a bit cleaner.

**Comment by LKY-stephen** [comment]
> > I guess a remaining question is the API for Frobenius (it's commented out in this branch currently). Would it make sense to have a fn frobenius() in Field itself? I think arkworks does that.
> 
> I'm not sure we would need an explicit Frobenius. I think the only places where it is used in Plonky2 is to compute inverses in OEFs, but this PR gives alternative implementations that doesn't use the Frobenius.
> 
> On the other hand, it would be pretty easy to implement it since it's the identity for prime fields. Also, since the Frobenius in the complex extension is the conjugate, we could remove the `conjugate()` function of `Mersenne31Complex` which could be a bit cleaner.

Thanks for reviewing. There is another reason I did not implement the frobenuis. If we check the frobenuis inverse algorithm, we can find the computation of r is based on the order of the tower field divided by the base field. In our case, this means for Mersenne31 complex's tower field, the r-value is p^2+1 for the quadratic field and p^4+p^2+1 for the cubic field, this difference makes a universal implementation difficult given we only expose the degree of the last irreducible poly(2,3). This is the reason I suggest only implement frobenuis staff for prime field and we can use this trait in inverse() is implemented.

File: baby-bear/src/baby_bear.rs [hunk 9]
```diff
@@ -384,17 +382,8 @@ mod tests {
...
-    fn two_adic_subgroup_zerofier() {
-        test_two_adic_subgroup_zerofier::<BabyBear>();
-    }
-
-    #[test]
-    fn two_adic_coset_zerofier() {
-        test_two_adic_coset_zerofier::<BabyBear>();
+    fn common_tests() {
+        field_test::<BabyBear>();
+        two_adic_field_test::<BabyBear>();
```
[the remark this answers is not in the document]
Comment (LKY-stephen) [reply baby-bear/src/baby_bear.rs]: Changed to macro so we can execute with better granularity.

**Comment by unzvfu** [comment]
@LKY-stephen Hi. One observation to start: Please don't force-push to the branch. This is almost never necessary and should absolutely be avoided on public repos since it breaks the repo on others' machines. Much better to just have the extra commits rather than rewrite the history.

**Comment by LKY-stephen** [comment]
> @LKY-stephen Hi. One observation to start: Please don't force-push to the branch. This is almost never necessary and should absolutely be avoided on public repos since it breaks the repo on others' machines. Much better to just have the extra commits rather than rewrite the history.

Oh, sorry. I wanted to keep the commit history clear and meaningful after merging. I will do this just before merging.

**Comment by unzvfu** [comment]
> > @LKY-stephen Hi. One observation to start: Please don't force-push to the branch. This is almost never necessary and should absolutely be avoided on public repos since it breaks the repo on others' machines. Much better to just have the extra commits rather than rewrite the history.
> 
> Oh, sorry. I wanted to keep the commit history clear and meaningful after merging. I will do this just before merging.

PRs are (or should be) squashed before merging anyway, so the clean commit history is not that important.

**Comment by LKY-stephen** [comment]
> > > @LKY-stephen Hi. One observation to start: Please don't force-push to the branch. This is almost never necessary and should absolutely be avoided on public repos since it breaks the repo on others' machines. Much better to just have the extra commits rather than rewrite the history.
> > 
> > 
> > Oh, sorry. I wanted to keep the commit history clear and meaningful after merging. I will do this just before merging.
> 
> PRs are (or should be) squashed before merging anyway, so the clean commit history is not that important.

If this repo prefers to squash and merge style then it is not necessary.

**Review by unzvfu — CHANGES_REQUESTED** [review]
Thanks for your contribution @LKY-stephen, much appreciated!

I've made a PR, #95, that captures most of the fixes I would have asked for (I wanted to check that what I wanted was reasonable, and the best way to do that was to implement them); you can look over the PR and merge (with whatever adjustments might be necessary). In summary:
1. Removed the redundant OEF trait
2. Fixed the spelling of "`ext_multiplicate_generator`"
3. Removed the redundant intermediate `scalar_mul` function
4. Removed generic implementation of "scalar multiplication" and provide specialisations

There are two other issues I'd like to raise which I didn't include in my "comment PR" since they require more discussion, and perhaps input from @dlubarov and/or @wborgeaud:
1. The extensions defined in this PR are not, strictly speaking, OEFs (they weren't really in Plonky2 either). Rather they are models for extensions defined via the data of an _irreducible binomial_, but they are more general than what is required to be an OEF (which restricts for example the structure of `p` and only being defined over a prime field, etc.). This is *good*, we want the more general version, but I think this is a good opportunity to use a more appropriate name than was used in Plonky2. I'm open to ideas, but as a first suggestion, I would replace the `OptimallyExtensible<D>` trait with `HasIrreducibleBinomial<D>` and the `QuadraticOef` (resp. `CubicOef`) struct with `BinomialQuadraticExtension` (resp. `BinomialCubicExtension`). (In #95 I removed the `OEF` trait, as it serves only to duplicate the data already in the `OptimallyExtensible` trait, so that's one less new name we need.)
2. There is a bigger issue here which is not introduced/caused by this particular PR, but is related to it: There is the potential for subtle bugs in the extension field code caused by glossing over the fact that representations of extension fields are not canonical. I don't think this PR is the right place to address this, but I do think anyone involved in extension field code should be aware of it. Please see #96 for more details on this issue.

**Comment by LKY-stephen** [comment]
@unzvfu, thanks for reviewing and suggesting improvements. I have merged them. Some outdated comments are also updated, together with some sanity checks about `from_base_slice`.

>I think this is a good opportunity to use a more appropriate name than was used in Plonky2. 

I also feel OEF is misleading especially since the first instance is based on MersenComplex, which is not a prime field. @dlubarov also mentioned the name issue. Given that `OEF` has gone, if we directly extend from the base field's trait, Maybe we can have two layers' trait? one base `Extendable`(which can help us support non-binomial cases in the future according to the previous comment) and one `BinomialExtendable` (which is for this PR). We can have other `XXExtenddable` for other cases if needed. I am OK with other suggested names.

>There is the potential for subtle bugs in the extension field code caused by glossing over the fact that representations of extension fields are not canonical
For this issue, I think transporting between different machines is not the purpose of the mentioned trait. Such a task needs a more rigorous serialization/deserialization scheme to handle. The `*_base_slice` functions are more like a way to assemble or dissemble the extension field for purposes like sampling or hashes.

**Comment by unzvfu** [comment]
> @unzvfu, thanks for reviewing and suggesting improvements. I have merged them. I also add a commit for updating the outdated comments and a missing sanity check about `from_base_slice`. I forced pushed this commit a little bit for handling the CI fmt error, hope this will not bother anyone.

Thank you! This is good to go now I think.

> > I think this is a good opportunity to use a more appropriate name than was used in Plonky2.
> 
> I also feel OEF is misleading especially since the first instance is based on MersenComplex, which is not a prime field. @dlubarov also mentioned the name issue. Given that `OEF` has gone, if we directly extend from the base field's trait, Maybe we can have two layers' trait? one base `Extendable`(which can help us support non-binomial cases in the future according to the previous comment) and one `BinomialExtendable` (which is for this PR). We can have other `XXExtenddable` for other cases if needed. I am OK with other suggested names.

Let's discuss this further in a separate issue/PR. I've opened #101 for that purpose.

> > There is the potential for subtle bugs in the extension field code caused by glossing over the fact that representations of extension fields are not canonical
> 
> For this issue, I feel communicating between different machines is not the purpose of the mentioned trait. Such a task needs a more rigorous serialization/deserialization scheme to handle. The `*_base_slice` functions are more like a way to assemble or dissemble the extension field for purposes like sampling or hashes.

Agreed. We are deferring addressing this issue for now, as discussed in #96.

**Review by unzvfu — APPROVED** [review]

**Pull request merged**